src/sentry/static/sentry/app/components/search/sources/index.jsx

```diff
@@ -0,0 +1,68 @@
+import {flatten} from 'lodash';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import {loadSearchMap} from '../../../actionCreators/formSearch';
+import ApiSource from './apiSource';
+import FormSource from './formSource';
+import RouteSource from './routeSource';
+
+class SearchSources extends React.Component {
+  static propTypes = {
+    query: PropTypes.string,
+    /**
+     * Render function the passes:
+     *
+     * `isLoading`
+     * `results` - Array of results
+     * `hasAnyResults` - if any results were found
+     */
+    children: PropTypes.func,
+  };
+
+  componentDidMount() {
+    // Loads form fields
+    loadSearchMap();
+  }
+
+  render() {
+    let {children, ...props} = this.props;
+
+    return (
+      <ApiSource {...props}>
+        {apiArgs => (
+          <FormSource {...props}>
+            {formFieldArgs => (
+              <RouteSource {...props}>
+                {routeArgs => {
+                  let allArgs = [apiArgs, formFieldArgs, routeArgs];
+                  // loading means if any result has `isLoading` OR any result is null
+                  let isLoading = !!allArgs.find(
+                    arg => arg.isLoading || arg.results === null
+                  );
+
+                  // Only use first `MAX_RESULTS` after sorting by score
+                  let foundResults =
+                    (!isLoading &&
+                      flatten(allArgs.map(({results}) => results || [])).sort(
```

Comment:
maybe just flatmap here?

Reply:
would be nice to pull this out of the render method for the sake of future reading and extension

Reply:
hmm #feelings - `flatten` + map seems easier to understand if you're not too familiar with functional concepts